File: app-geekhome/src/main/java/eu/geekhome/rest/plugins/PluginDtoMapper.java
package eu.geekhome.rest.plugins;

import eu.geekhome.services.extensibility.PluginMetadata;
import eu.geekhome.services.hardware.HardwarePlugin;
import org.pf4j.Plugin;
import org.pf4j.PluginState;
import org.pf4j.PluginWrapper;

public class PluginDtoMapper {

    public PluginDto map(PluginWrapper pluginWrapper) {
        Plugin plugin = pluginWrapper.getPlugin();
        String id = pluginWrapper.getPluginId();
        String version = pluginWrapper.getDescriptor().getVersion();
        String provider = pluginWrapper.getDescriptor().getProvider();
        boolean enabled = pluginWrapper.getPluginState() == PluginState.STARTED;
        boolean isHardwareFactory = pluginWrapper.getPlugin() instanceof HardwarePlugin;

        if (plugin instanceof PluginMetadata) {
            PluginMetadata metadata = (PluginMetadata)plugin;
            return new PluginDto(id, metadata.getName(), metadata.getDescription(), provider, version, isHardwareFactory, enabled);
        } else {
            return new PluginDto(id,null, null, provider, version, isHardwareFactory, enabled);
        }
    }
}

File: app-geekhome/src/main/java/eu/geekhome/rest/plugins/PluginDtoMapper.kt
package eu.geekhome.rest.plugins

import eu.geekhome.R
import eu.geekhome.rest.settinggroup.SettingGroupDtoMapper
import org.pf4j.PluginWrapper
import org.pf4j.PluginState
import eu.geekhome.domain.hardware.HardwarePlugin
import eu.geekhome.domain.extensibility.PluginMetadata
import javax.inject.Inject

class PluginDtoMapper @Inject constructor(
    private val settingsGroupDtoMapper: SettingGroupDtoMapper,
) {
    fun map(pluginWrapper: PluginWrapper): PluginDto {
        val plugin = pluginWrapper.plugin
        val id = pluginWrapper.pluginId
        val version = pluginWrapper.descriptor.version
        val provider = pluginWrapper.descriptor.provider
        val enabled = pluginWrapper.pluginState == PluginState.STARTED
        val isHardwareFactory = pluginWrapper.plugin is HardwarePlugin
        return if (plugin is PluginMetadata) {
            val metadata = plugin as PluginMetadata
            val settingGroups = metadata.settingGroups.map { settingsGroupDtoMapper.map(it) }
            PluginDto(id, metadata.name, metadata.description, provider, version, isHardwareFactory, enabled, settingGroups)
        } else {
            PluginDto(id, R.plugin_no_name, R.plugin_no_description, provider, version, isHardwareFactory, enabled, null)
        }
    }
}